File: src/model/steps/StepManager.js
const StepTypes = require('./StepTypes')
const ProcessStep = require('./ProcessStep')
const DecisionStep = require('./DecisionStep')
const SendEmailStep = require('./SendEmailStep')
const CsvReaderStep = require('./CsvReaderStep')
const HtmlParserStep = require('./GetHtmlStep')
const PromptStep = require('./PromptStep')
const ExecStep = require('./ExecStep')

class StepManager {

    constructor() {
        this._steps = {}
    }

    createInstance(type) {
        const stepClass = this._steps[type]
        if (!stepClass) throw new Error(`類別 ${type} 不存在!`)
        return new stepClass()
    }

    init() {
        this.registerStepClass(StepTypes.START, ProcessStep)
        this.registerStepClass(StepTypes.STOP, ProcessStep)
        this.registerStepClass(StepTypes.PROCESS, ProcessStep)
        this.registerStepClass(StepTypes.DECISION, DecisionStep)
        this.registerStepClass(StepTypes.SEND_EMAIL, SendEmailStep)
        this.registerStepClass(StepTypes.CSV_READER, CsvReaderStep)
        this.registerStepClass(StepTypes.GET_HTML, HtmlParserStep)
        this.registerStepClass(StepTypes.PROMPT, PromptStep)
        this.registerStepClass(StepTypes.EXEC, ExecStep)
    }

    registerStepClass(type, stepClass) {
        this._steps[type] = stepClass
    }
}

module.exports = new StepManager()

File: Models/Steps/StepManager.cs
namespace WorkflowEngine.Models.Steps;

public static class StepManager
{
    private static readonly Dictionary<string, Type> StepTypes = new();

    public static void Init()
    {
        RegisterStep(Steps.StepTypes.START, typeof(StartStep));
        RegisterStep(Steps.StepTypes.STOP, typeof(StopStep));
        RegisterStep(Steps.StepTypes.PROCESS, typeof(ProcessStep));
        RegisterStep(Steps.StepTypes.DECISION, typeof(DecisionStep));
        RegisterStep(Steps.StepTypes.PROMPT, typeof(PromptStep));
        RegisterStep(Steps.StepTypes.SEND_EMAIL, typeof(SendEmailStep));
        RegisterStep(Steps.StepTypes.CSV_READER, typeof(CsvReaderStep));
        RegisterStep(Steps.StepTypes.GET_HTML, typeof(GetHtmlStep));
        RegisterStep(Steps.StepTypes.EXEC, typeof(ExecStep));
    }

    public static void RegisterStep(string typeName, Type stepType)
    {
        StepTypes[typeName] = stepType;
    }

    public static Step CreateStep(string typeName)
    {
        if (StepTypes.TryGetValue(typeName, out var stepType))
        {
            return (Step)(Activator.CreateInstance(stepType) ?? throw new Exception($"無法建立步驟類型: {typeName}"));
        }

        throw new Exception($"未知的步驟類型: {typeName}");
    }
}